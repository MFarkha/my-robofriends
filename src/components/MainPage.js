import { Component } from 'react';
import CardList from './CardList';
import SearchBox from './SearchBox';
import './MainPage.css'
import Scroll from './Scroll'
import ErrorBoundary from './ErrorBoundary';
import Header from './Header';

class MainPage extends Component {
    componentDidMount() {
        this.props.onRequestRobots();
    }
    filterRobots = () => {
        const { robots, searchField } = this.props;
        try {
            return robots.filter(robot => {
                return robot.name.toLowerCase().includes(searchField.toLowerCase());
            });
        } catch(err) {
            this.setState({error: err});
            return [];
        }
    }
    render() {
        const { onSearchChange, isPending } = this.props;
        if (isPending) {
            return <h1>Loading</h1>;
        } else {
            return (
            <div className='tc'>
                <Header/>
                <SearchBox searchChange={onSearchChange}/>
                <Scroll>
                    <ErrorBoundary>
                        <CardList robots={ this.filterRobots() }/>
                    </ErrorBoundary>
                </Scroll>
            </div>
            );
        }
    }
}

export default MainPage;