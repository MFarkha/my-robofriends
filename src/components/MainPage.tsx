import { Component, ChangeEventHandler } from 'react';
import CardList from './CardList';
import SearchBox from './SearchBox';
import './MainPage.css'
import Scroll from './Scroll'
import ErrorBoundary from './ErrorBoundary';
import Header from './Header';
import { AppState, IRobot } from '../app/store';


interface IMainPageProps extends AppState {
    onRequestRobots: () => void,
    onSearchChange: ChangeEventHandler<HTMLInputElement>,
};

class MainPage extends Component<IMainPageProps> {
    componentDidMount() {
        this.props.onRequestRobots();
    }
    filterRobots = () => {
        const { robots, searchField, error } = this.props;
        if (error.length!==0) {
            return []
        } else {
            return (robots as Array<IRobot>).filter(robot => (
                robot.name.toLowerCase().includes((searchField as string).toLowerCase())
            ))
        }
    }
    render() {
        const { onSearchChange, isPending, error } = this.props;
        if (isPending) {
            return <h1>Loading</h1>;
        } else {
            return (
            <div className='tc'>
                <Header/>
                <SearchBox searchChange={ onSearchChange }/>
                <Scroll>
                    <ErrorBoundary>
                        <CardList error={ error } robots={ this.filterRobots() }/>
                    </ErrorBoundary>
                </Scroll>
            </div>
            );
        }
    }
}

export default MainPage;