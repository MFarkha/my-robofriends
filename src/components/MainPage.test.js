import React from 'react';
import { shallow } from 'enzyme';
import MainPage from './MainPage';

var wrap;

beforeEach(() => {
    const mockProps = {
        robots: [],
        searchField: '',
        isPending: false,
        error: '',
        onRequestRobots: jest.fn()
    }
    wrap = shallow(<MainPage {...mockProps }/>);
});

it('expects to render the MainPage component without crashing', () => {
    expect(wrap).toMatchSnapshot();
});

it('filters the robots correctly with an empty searchField', () => {
    expect(wrap.instance().filterRobots([], '')).toEqual([]);
});

it('filters the robots correctly with a searchField', () => {
    const mockProps = {
        robots: [{
            id: 1,
            name: 'John Snow',
            email: 'john@example.com'
        }],
        searchField: 'Marry',
        isPending: false,
        error: '',
        onRequestRobots: jest.fn()
    }
    wrap = shallow(<MainPage {...mockProps }/>);

    expect(wrap.instance().filterRobots()).toEqual([]);
});

it('filters the robots correctly with a searchField 2', () => {
    const mockProps = {
        robots: [{
            id: 1,
            name: 'John Snow',
            email: 'john@example.com'
        }],
        searchField: 'John',
        isPending: false,
        error: '',
        onRequestRobots: jest.fn()
    }
    wrap = shallow(<MainPage {...mockProps }/>);

    expect(wrap.instance().filterRobots()).toEqual([{
        id: 1,
        name: 'John Snow',
        email: 'john@example.com'
    }]);
});
it('sets a props.error in case of undefined value of robots', () => {
    const mockProps = {
        robots: undefined,
        searchField: '',
        isPending: false,
        error: '',
        onRequestRobots: jest.fn()
    }
    wrap = shallow(<MainPage {...mockProps }/>);

    expect(wrap.instance().filterRobots()).toEqual([]);
    expect(wrap.props.error).not.toEqual('');
});
it('displays a `loading..` message when `robots` data are pending', () => {
    const mockProps = {
        robots: [],
        searchField: '',
        isPending: true,
        error: '',
        onRequestRobots: jest.fn()
    }
    wrap = shallow(<MainPage {...mockProps }/>);

    expect(wrap).toMatchSnapshot();
});

