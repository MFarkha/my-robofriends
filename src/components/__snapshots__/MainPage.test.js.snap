// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays a \`loading..\` message when \`robots\` data are pending 1`] = `
<h1>
  Loading
</h1>
`;

exports[`expects to render the MainPage component without crashing 1`] = `
<div
  className="tc"
>
  <Header />
  <SearchBox />
  <Scroll>
    <ErrorBoundary>
      <CardList
        robots={Array []}
      />
    </ErrorBoundary>
  </Scroll>
</div>
`;
